src/parser/priest/discipline/modules/features/AtonementApplicatorBreakdown.js

```diff
@@ -0,0 +1,256 @@
+    _hasDepth = false;
+
+    constructor(...args) {
+        super(...args);
+
+        this._hasEL = this.selectedCombatant.hasTrait(SPELLS.ENDURING_LUMINESCENCE.id);
+        this._hasDepth = this.selectedCombatant.hasTrait(SPELLS.DEPTH_OF_THE_SHADOWS.id);
+
+        this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(SPELLS.EVANGELISM_TALENT), this.handleEvangelismCasts);
+        this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(SPELLS.SHADOW_MEND), this.storeShadowMendsCasts);
+        this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(SPELLS.POWER_WORD_SHIELD), this.storePowerWordShieldsCasts);
+        this.addEventListener(Events.applybuff.by(SELECTED_PLAYER).spell(SPELLS.ATONEMENT_BUFF), this.assignAtonementBuffToApplicator);
+        this.addEventListener(Events.refreshbuff.by(SELECTED_PLAYER).spell(SPELLS.ATONEMENT_BUFF), this.assignAtonementBuffToApplicator);
+        this.addEventListener(Events.heal.by(SELECTED_PLAYER).spell(SPELLS.ATONEMENT_HEAL_NON_CRIT), this.handleAtonementsHits);
+        this.addEventListener(Events.heal.by(SELECTED_PLAYER).spell(SPELLS.ATONEMENT_HEAL_CRIT), this.handleAtonementsHits);
+    }
+
+    storeShadowMendsCasts(event){
+        this._castsAplyBuffsMap.set({
+            "event": event,
+            "applicatorId": SPELLS.SHADOW_MEND.id,
+        }, null);
+    }
+    
+    storePowerWordShieldsCasts(event){
+        this._castsAplyBuffsMap.set({
+            "event": event,
+            "applicatorId": SPELLS.POWER_WORD_SHIELD.id,
+        }, null);
+
+    }
+
+    assignAtonementBuffToApplicator(event){
+        if(event.__modified === true){ // Power Word: Radiance
+
+            //Set the wasRefreshed property of the old atonement on the same target to true
+            //so we can stop attributing atonement healing to the old atonement
+            if(event.type === "refreshbuff"){
+                this.setWasRefreshedProperty(event, true);
+            }
+
+            //Putting a custom event object for Radiances since there it's only 1 cast for 5 buffs
+            this._castsAplyBuffsMap.set({
+                "event": {
+                    "targetID": event.targetID,
+                },
+                "applicatorId": SPELLS.POWER_WORD_RADIANCE.id,
+            }, 
+            {
+                "applyBuff": event,
+                "atonementEvents": [],
+                "extendedByEvangelism": false,
+                "wasRefreshed": false,
+            });
+        } else { //Shadow Mend and Power Word: Shield
+            if(event.type === "refreshbuff"){
+                this.setWasRefreshedProperty(event, true);
+            }
+
+            //Iterate the casts in reverse to break out at the most recent appearance of the corresponding targetID
+            const playerWithAtonement = event.targetID;
+            const reversedMapKeys = Array.from(this._castsAplyBuffsMap.keys()).slice().reverse();
+
+            for(const cast of reversedMapKeys) {
```

Comment:
I think writing this with functional array functions would be cleaner. I recommend avoiding for-loops as they can be a lot more troublesome.

```
const mostRecentCastApplyBuff = reversedMapKeys.find(cast => cast.event && cast.event.targetID === playerWithAtonement)
if (mostRecentCastApplyBuff) {
  this._castsAplyBuffsMap.set(mostRecentCastApplyBuff, {
                            "applyBuff": event,
                            "atonementEvents": [],
                            "extendedByEvangelism": false,
                            "wasRefreshed": false,
                        });
}
```